### DISCUSSION TITLE
Pickling issue for the Stumpi

### DISCUSSION THREAD

**Discussion by soja-soja** [TDAmeritrade/stumpy#710] [Help: Coding & Implementations]
Hi, 

in case of ongoing streaming of data in different groups (so dealing with more than one stumpi instance), one efficient approach might be to pickle the instances, and then use them whenever new datapoints related to that group is available. 

However, due to the implementation I believe the class is not pickleable. Any thought on that? 

replicating that would be as easy as : 

```
import pickle
import stumpy

x = stumpy.stumpi(list(map(float, [1,2,3,4,5,6,7,8])), m=3)
pickle.dumps(x)
```

**Comment by seanlaw** [comment]
@soja-soja Thank you for your question and welcome to the STUMPY community. Unfortunately, functions like `stumpy.stumpi` are "decorated" functions and it's not easy/possible to pickle them (see [this past attempt](https://github.com/TDAmeritrade/stumpy/issues/488)). Would it be possible to use the more flexible `cloudpickle` package instead?

```
import numpy as np
import stumpy
import cloudpickle


if __name__ == "__main__":
    T = np.random.rand(100)
    m = 50

    stream = stumpy.stumpi(T, m, egress=True)
    stream.update(np.random.rand())
    print(stream.left_P_)

    dump_stream = cloudpickle.dumps(stream)

    stream = cloudpickle.loads(dump_stream)
    print(stream.left_P_)
```

**Reply by NimaSarajpoor** [reply]
@seanlaw 
Is issue #621 related? Because I think `stumpi` should be fast if we allow user pass `mp` (of course, we may want to save some other stuff as well like sliding mean and std). one can save `mp` (as `mp_file.npy`) and then just load it and pass it to the stumpi.

**Reply by soja-soja** [reply]
Thanks for the prompt response, that seems like a good alternative, hopefully I can have cloudpickle in my project. 

Also per Nima's suggestion, having  pre-computed `mp` is a great idea and this one is also a use-case for that. I think we cannot rely solely on the current use-cases and I'm sure more long-term use-cases will be using stumpy in the future.
